Hide 'align' instruction behind jmp

kunalspathak opened the pull request:
### Overview

With current loop alignment, `align` instruction are placed just before the loop start. This can sometimes affect performance adversely if the block in which `align` instruction is placed is hot or if it is part of nested loop, in which case processor would perform fetching and decoding of `nop`. This PR places the `align` instructions behind unconditional `jmp` instructions if they exist before the loop that it is trying to align. If no such `jmp` were present, then `align` will be placed right before the loop start, the one it is done today.

Here is the sample diff where `align` instruction was moved from `IG31` and placed into `IG29` after the `jmp` instruction.

![image](https://user-images.githubusercontent.com/12488060/138524839-cfa5d007-ae43-473a-baf1-f1a6d2d75e30.png)

I have also added `COMPlus_JitHideAlignBehindJmp` to turn off this feature for debugging purpose. In Release, it is always ON. I have also added a stress mode where 50% of time, we would emit `INS_BREAKPOINT` instead of `align` in situation where the `align` are placed behind `jmp` to make sure that we never execute them.

### Design

A new datastructure `alignBlocksList` is created which is a linked list of all `BasicBlock` that are the head of loop start needing alignment. This is done during final ref counting phase. During codegen, we pull the basic block from the `alignBlocksList` and monitor if there is any unconditional jmp. If we find one, we emit `align` instruction and set the BB flag `BBF_LOOP_ALIGN_ADDED`. This makes sure that if we see more `jmp` before the actual loop start, we do not add `align` instructions again. When we reach a point in the flow where next block is the loop start, we would update the `targetIG` (see below) of the `alignInstr`. At this time, we also determine that if we didn't see any `jmp` so far (using `BBF_LOOP_ALIGN_ADDED`), we would emit the `align` instruction. Finally, we move to the next `BasicBlock` of `alignBlocksList`.

`instrDescAlign` data structure has been updated. `idaIG` field in it now points to the IG that contains the `align` instruction. This IG can be before the loop or can be some on the previous IG that ends with `jmp`. `idaTargetIG` is the IG that earlier used to be `idaIG`. It points to the IG before the IG that has loop. This is used when we want to calculate the `loopSize`. Some of the changes are around getting the right field wherever necessary.

`IGF_LOOP_ALIGN` flag, which previously used to be on an IG just before the loop IG has been replaced by `IGF_HAS_ALIGN` and will be on IG that contains the `align` instruction. Again, this may or may not be the one just before the loop IG. Finally, to handle special scenarios where an IG that is part of loop might have `align` instruction for a different IG, flag `IGF_REMOVED_ALIGN` is added that tells if the `align` instruction present in that IG are removed or not.

### Impact

Ideally, this change should have been no code size diffs, however the reason behind diffs on x64 is that because of moving the `align` instruction, we tend to shorten some jumps and that changes the heuristics calculation of alignment. However, as seen below, the impact is minimal and number of unchanged methods are far more than the impacting methods. 

As expected, there is no code size difference in arm64 because we just moved the `align` instruction around.


| collection        | platform      | main   | PR     | diff | diff % | methods regressed | methods improved | methods unchanged |
|-------------------|---------------|--------|--------|------|--------|-------------------|------------------|-------------------|
| libraries.pmi     | windows.x64   | 577335 | 577170 | -165 | -0.03% | 27                | 40               | 546               |
| benchmarks.run    | windows.x64   | 277173 | 277211 | 38   | 0.01%  | 19                | 13               | 181               |
| coreclr_tests.pmi | windows.x64   | 188577 | 188598 | 21   | 0.01%  | 9                 | 7                | 184               |
| aspnet.run        | windows.x64   | 113942 | 113995 | 53   | 0.05%  | 3                 | 3                | 66                |
| benchmarks.run    | windows.arm64 | 201028 | 201028 | 0    | 0.00%  | 0                 | 0                | 132               |
| coreclr_tests.pmi | windows.arm64 | 118012 | 118012 | 0    | 0.00%  | 0                 | 0                | 234               |
| libraries.pmi     | windows.arm64 | 372600 | 372600 | 0    | 0.00%  | 0                 | 0                | 405               |

Detail diffs: https://gist.github.com/kunalspathak/9cc028b60a2e7aba82308fa1e94951ba

Contributes to https://github.com/dotnet/runtime/issues/43227

kunalspathak committed:
Hide align behind a jmp

fix the alignBytesRemoved

Some fixes and working model

Some fixes and redesign

Some more fixes

more fixes

fix

Add the check  for fgFirstBB

misc changes

code cleanup + JitHideAlignBehindJmp switch

validatePadding only if align are before the loop IG

More cleanup, remove commented code

jit format

kunalspathak committed:
Add stress mode to emit int3 for xarch

kunalspathak committed:
Fix a problem where curIG==0 and loop might be emitted in curIG, adjust the targetIG to prevIG

Add IGF_REMOVED_ALIGN flag for special scenarios

kunalspathak committed:
Add stress mode to emit bkpt for arm64

BruceForstall reviewed (changes requested):
Some additional questions/comments:

- Is there a measured perf difference? Or is this speculative, and we'll see what the lab shows?
- What if there are many "jmp" before the loop to align? Maybe the first one is a poor choice because it is in a hot loop. Or maybe the only one prior is in a hot loop. Maybe we shouldn't move it then, due to impacting the I-cache? Should we look at block weights to decide?
- It seems like there should be an "alignment planning" pass that decides where to put the alignment instructions, and annotates the BasicBlocks with those decisions. It could both set a flag that an alignment instruction is needed, and include a pointer to the BasicBlock for the loop we are aligning. Then, the codegen loop would just act on those decisions, and be very simple. Interleaving the planning and codegen loops seems complicated. It seems like this might remove the need for the `alignBBLists` list.
- Is there an end-to-end design written down in comments in one place? It seems like there should be.

kunalspathak committed:
Add a loop align instruction placement phase

kunalspathak commented on src/coreclr/jit/emit.h:
```diff
@@ -1381,13 +1391,18 @@ class emitter
+        instrDescAlign* idaNext;     // next align in the group/method
+        insGroup*       idaIG;       // containing group
+        insGroup*       idaTargetIG; // The IG before the loop IG.
+                                     // If no 'jmp' instructions were found until targetIG,
+                                     // the 'align' will be placed in this IG.
```
> I'm confused by this description. For one, this is an instrDesc of an align instruct that is already in an IG, so "will be placed in this IG" doesn't make sense.

I think better way to put is: "If no 'jmp' instructions were found until targetIG, then `idaTargetIG` will be same as `idaIG`." Let me know if that makes sense.


> Does the align instruction get placed in a JMP IG, with idaTargetIG pointing the the IG before the loop -- the "last possible place the alignment can be placed", and then we might later generate an align instruction into a subsequent JMP IG that still precedes the target?

This logic has been changed now because of separate "loop alignment placement" phase. The IG, either the one that ends with `jmp` or the one before the loop, will be determine in before code generation and the `align` will be placed in the corresponding IG. 

For scenarios where `align` is placed in a different IG (`IG03`) that doesn't precede the loop that starts at `IG06`, `idaTargetIG == IG05` and `idaIG == IG03`.

```asm
IG03:
      ...
      jmp IG19
                ; <---- align instruction placement
IG04:
    ...

IG05
   ...
   jne IG08

IG06:
  ...
  ...

IG07:
  ...
   jne IG06
  
```


For scenarios where `align` is placed in a immediately before the IG with loop that starts at `IG05`, `idaTargetIG == idaIG == IG04`.


```asm
IG04:
    ...
    jne IG08
                ; <---- align instruction placement
IG05
   ...
   ...

IG06:
  ...
  jne IG05

```

`idaTargetIG` will be point to the if `align` was placed in non-preceding IG

> The igTargetIG points forwards in the code stream, right? So is it null for a while, then filled in?

Correct. It is filled in `emitConnectAlignInstrWithCurIG()`.

BruceForstall commented on src/coreclr/jit/lclvars.cpp:
```diff
@@ -4527,13 +4529,41 @@ void Compiler::lvaComputeRefCounts(bool isRecompute, bool setSlotNumbers)
     {
         if (block->IsLIR())
         {
+#if FEATURE_LOOP_ALIGN
+            if (calculateAlign)
+            {
+                // Track the blocks that needs alignment
+                // except if it is first block, because currently adding padding
+                // in prolog is not supported
+                if (opts.compJitHideAlignBehindJmp && block->isLoopAlign() && (block != fgFirstBB))
+                {
+                    alignBlocksList* curr = new (this, CMK_FlowList) alignBlocksList(block);
+
+                    if (alignBBLists == nullptr)
+                    {
+                        alignBBLists = curr;
+                    }
+                    else
+                    {
+                        alignBB->next = curr;
+                    }
+
+                    alignBB = curr;
+                }
+            }
+#endif
+
```
Could you keep a count of aligned loops and decrement the count when you remove a BBF_LOOP_ALIGN bit, then just check for "loopAlignCount > 0" before running the PlaceLoopAlignment phase?

It's really unpleasant to have unrelated phases tied together in a somewhat implicit contract.

kunalspathak replied:
> Could you keep a count of aligned loops and decrement the count when you remove a BBF_LOOP_ALIGN bit, then just check for "loopAlignCount > 0" before running the PlaceLoopAlignment phase?

Yes, that's exactly what I tried doing, but turns out that sometimes a block/loop is marked as not needing alignment multiple times specially in `AddContainsCallAllContainingLoops()` that miscalculates the count.

BruceForstall replied:
> in AddContainsCallAllContainingLoops() that miscalculates the count.

Why would it do that? Once you've cleared the bit, you wouldn't decrement again. e.g.,
```
void Compiler::AddContainsCallAllContainingLoops(unsigned lnum)
{

#if FEATURE_LOOP_ALIGN
    // If this is the inner most loop, reset the LOOP_ALIGN flag
    // because a loop having call will not likely to benefit from
    // alignment
    if (optLoopTable[lnum].lpChild == BasicBlock::NOT_IN_LOOP)
    {
        BasicBlock* first = optLoopTable[lnum].lpFirst;
        if (first->isLoopAlign())
        {
            assert(compAlignedLoopCount > 0);
            --compAlignedLoopCount;
            first->bbFlags &= ~BBF_LOOP_ALIGN;
            JITDUMP("Removing LOOP_ALIGN flag for " FMT_LP " that starts at " FMT_BB " because loop has a call.\n", lnum,
                first->bbNum);
         }
    }
#endif
...
```

kunalspathak committed:
review comments

kunalspathak committed:
Change from unsigned short to unsigned

kunalspathak commented on src/coreclr/jit/emit.h:
```diff
@@ -1777,15 +1795,26 @@ class emitter
+
+    // Points to the most recent added align instruction. If there are multiple align instructions like in arm64 or
+    // non-adpative alignment on xarch, this points to the first align instruction of the series of align instructions.
+    instrDescAlign* emitRecentFirstAlign;
```
> Is this the same as `emitAlignLast`, except if there are multiple consecutive align instructions used?

Exactly.

> How about the name `emitAlignLastGroup` instead?

Sure.

kunalspathak commented on src/coreclr/jit/emit.cpp:
```diff
@@ -5492,6 +5556,28 @@ unsigned emitter::emitCalculatePaddingForLoopAlignment(insGroup* ig, size_t offs
+//
+emitter::instrDescAlign* emitter::emitAlignInNextIG(instrDescAlign* alignInstr)
+{
+#if defined(TARGET_XARCH)
+    if (!emitComp->opts.compJitAlignLoopAdaptive)
+#endif
```
true.

kunalspathak commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -9815,32 +9822,74 @@ static BYTE* emitOutputNOP(BYTE* dstRW, size_t nBytes)
+        // to the nop instruction in disasm.
+        // e.g. CC                   align    [1 bytes for IG29]
+        //
+        //if (emitComp->opts.disAsm)
+        //{
+        //    emitDispInsAddr(dstRW);
+
+        //    emitDispInsOffs(0, false);
+
+        //    printf("                      %-9s  ; stress-mode injected interrupt\n", "int3");
+        //}
```
I intentionally left it as commented so during debugging, we could uncomment and see the instruction. Let me know if you think otherwise.

kunalspathak committed:
review comments around cleanup

kunalspathak committed:
emitForceNewIG

kunalspathak committed:
Remove emitPrevIG

kunalspathak committed:
Revert change to forceNewIG for align instruction

kunalspathak committed:
Use loopAlignCandidates

kunalspathak committed:
Use loopHeadIG reference

kunalspathak committed:
jit format

kunalspathak committed:
Remove unneeded method

kunalspathak commented on src/coreclr/jit/codegen.h:
```diff
@@ -217,6 +217,10 @@ class CodeGen final : public CodeGenInterface
     void genInitializeRegisterState();
 
+#if FEATURE_LOOP_ALIGN
+    void genMarkBlocksForLoopAlignment();
```
Need to delete this.

kunalspathak committed:
Misc changes

kunalspathak commented:
@BruceForstall - This should be ready for the review. Here are the changes:
- Added `bbPlaceLoopAlignInstructions()` pass to label the blocks that should contain `align` instruction. It prefers cold blocks if there are multiple candidates. Also track `loopAlignCandidates` and skip this phase if `loopAlignCandidates == 0`. The code is much simpler now.
- Removed the need of `prevIG` that I was tracking for few edge scenarios and instead force a new IG before loop head.
- I have created `loopHeadIG()` that retrieves the loop head IG that we are interested in most often. I still need to track the previous IG (`idaTargetIG` has been renamed to `idaLoopHeadPredIG`) so I can know after what point I can re-enable the VEX encoding optimization. If we decide to re-enable the VEX encoding after loop head IG, then I noticed that the deoptimized version of VEX encoding is generated for instructions inside the loop.

kunalspathak commented:
CodeSize diffs: https://dev.azure.com/dnceng/public/_build/results?buildId=1464807&view=ms.vss-build-web.run-extensions-tab
PerfScore diffs:

| Name                                    | Diff     | Methods regressed | Methods improved | Methods unchanged |
|-----------------------------------------|----------|-------------------|------------------|-------------------|
| benchmarks.run.windows.x64.checked      | -6.69    | 28                | 90               | 100               |
| libraries.pmi.windows.x64.checked       | -142.01  | 59                | 242              | 319               |
| aspnet.run.windows.x64.checked          | -333.71  | 10                | 39               | 28                |
| coreclr_tests.pmi.windows.x64.checked   | -34.29   | 52                | 64               | 67                |
| benchmarks.run.windows.arm64.checked    | -2052.67 | 1                 | 38               | 83                |
| libraries.pmi.windows.arm64.checked     | -714.32  | 13                | 104              | 307               |
| coreclr_tests.pmi.windows.arm64.checked | 84.04    | 153               | 28               | 38                |

BruceForstall commented:
> I analyzed some of the PerfScore regressions on windows/arm64 for coreclr_tests, and they all are coming from moving the align instructions behind `jmp` and those blocks are expensive.

Should PerfScore not count align instructions in an IG following an unconditional branch?

BruceForstall commented on src/coreclr/jit/compiler.cpp:
```diff
@@ -5322,6 +5332,61 @@ void Compiler::compCompile(void** methodCodePtr, uint32_t* methodCodeSize, JitFl
+
+//------------------------------------------------------------------------
+// bbPlaceLoopAlignInstructions: Iterate over all the blocks and determine
+//      the best position to place the 'align' instruction. After 'jmp' are
+//      preferred over block before loop and in case there are multiple blocks
```
```suggestion
//      the best position to place the 'align' instruction. Inserting 'align' instructions after an unconditional branch is
//      preferred over inserting in the block before the loop. In case there are multiple blocks
```

BruceForstall commented on src/coreclr/jit/compiler.cpp:
```diff
@@ -5322,6 +5332,61 @@ void Compiler::compCompile(void** methodCodePtr, uint32_t* methodCodeSize, JitFl
+//      the best position to place the 'align' instruction. After 'jmp' are
+//      preferred over block before loop and in case there are multiple blocks
+//      having 'jmp', the one that has lower weight is preferred.
+//      If the block having 'jmp' is hot than the block before loop, the align
```
```suggestion
//      If the block having 'jmp' is hotter than the block before the loop, the align
```

BruceForstall commented on src/coreclr/jit/compiler.cpp:
```diff
@@ -5322,6 +5332,61 @@ void Compiler::compCompile(void** methodCodePtr, uint32_t* methodCodeSize, JitFl
+    assert(loopAlignCandidates > 0);
+
+    // Add align only if there were any loops that needed alignment
+    weight_t    minBlockSoFar = INFINITE;
```
Where is `INFINITE` coming from? I think you want `BB_MAX_WEIGHT` instead.

BruceForstall commented on src/coreclr/jit/compiler.cpp:
```diff
@@ -5322,6 +5332,61 @@ void Compiler::compCompile(void** methodCodePtr, uint32_t* methodCodeSize, JitFl
+            {
+                minBlockSoFar = block->bbWeight;
+                bbHavingAlign = block;
+                JITDUMP(FMT_BB ", bbWeight= %f ends with unconditional 'jmp' \n", block->bbNum, block->bbWeight);
```
```suggestion
                JITDUMP(FMT_BB ", bbWeight=" FMT_WT " ends with unconditional 'jmp' \n", block->bbNum, block->bbWeight);
```

BruceForstall commented on src/coreclr/jit/compiler.cpp:
```diff
@@ -5322,6 +5332,61 @@ void Compiler::compCompile(void** methodCodePtr, uint32_t* methodCodeSize, JitFl
+            }
+            else
+            {
+                JITDUMP("Marking " FMT_BB " that ends with uncondition jump with BBF_HAS_ALIGN for loop at " FMT_BB
```
```suggestion
                JITDUMP("Marking " FMT_BB " that ends with unconditional jump with BBF_HAS_ALIGN for loop at " FMT_BB
```

BruceForstall commented on src/coreclr/jit/compiler.cpp:
```diff
@@ -5322,6 +5332,61 @@ void Compiler::compCompile(void** methodCodePtr, uint32_t* methodCodeSize, JitFl
+            bbHavingAlign->bbFlags |= BBF_HAS_ALIGN;
+            minBlockSoFar = INFINITE;
+            bbHavingAlign = nullptr;
+        }
```
Can you early exit when there are no more? I.e., add:
```
if (--loopAlignCandidates == 0)
{
    break;
}
```

BruceForstall commented on src/coreclr/jit/compiler.cpp:
```diff
@@ -5266,6 +5268,14 @@ void Compiler::compCompile(void** methodCodePtr, uint32_t* methodCodeSize, JitFl
 #endif
 
+#if FEATURE_LOOP_ALIGN
+    if (loopAlignCandidates > 0)
```
IMO, this check should happen within the phase, and just early out. Makes the "list of phases" part of the code simpler.

BruceForstall commented on src/coreclr/jit/compiler.cpp:
```diff
@@ -5322,6 +5332,61 @@ void Compiler::compCompile(void** methodCodePtr, uint32_t* methodCodeSize, JitFl
+#if FEATURE_LOOP_ALIGN
+
+//------------------------------------------------------------------------
+// bbPlaceLoopAlignInstructions: Iterate over all the blocks and determine
```
IMO, this function doesn't need a `bb` prefix

BruceForstall commented on src/coreclr/jit/compiler.h:
```diff
@@ -8362,6 +8363,12 @@ class Compiler
                            weight_t refCntWtdStkDbl);
 #endif // DOUBLE_ALIGN
 
+#if FEATURE_LOOP_ALIGN
+public:
+// alignBlocksList* alignBBLists;
+
+#endif // FEATURE_LOOP_ALIGN
+
```
Remove

BruceForstall commented on src/coreclr/jit/emit.cpp:
```diff
@@ -913,6 +913,9 @@ insGroup* emitter::emitSavIG(bool emitAdd)
         emitAlignLast = last;
+        // Point to the first instruction of most recent
+        // align instruction(s) added.
+        emitAlignLastGroup = list;
```
It looks like the emitAlignList is in the reverse order of the emitCurIGAlignList loop, due to inserting in reverse order. So, does `list` actually represent the earliest-in-program-order `align` instruction, or the last one? I presume the code here wants it to be the "earliest". If that's indeed so, can you explain why?

BruceForstall replied:
Hmmm... It looks like `emitLoopAlign` creates the list in inverse-of-program-order, so this actually reverses that. Worth a comment.

kunalspathak replied:
Yes - It is essentially in the same way we create `emitCurIGjmpList`. I will add a comment.

BruceForstall commented on src/coreclr/jit/emit.cpp:
```diff
@@ -4818,10 +4821,10 @@ void emitter::emitCheckAlignFitInCurIG(unsigned short nAlignInstr)
-//  So insert an alignment instruction here to ensure that
-//  we can properly align the code.
+//  So insert an alignment instruction of "paddingBytes" to ensure that
+//  the code is properly aligned.
```
Explain the `isFirstAlign` argument here (preferably with the new-style function header comment)

BruceForstall commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -9924,32 +9931,74 @@ static BYTE* emitOutputNOP(BYTE* dstRW, size_t nBytes)
+        // it. As such, leave it as is, the "0xCC" bytecode will be seen next
+        // to the nop instruction in disasm.
+        // e.g. CC                   align    [1 bytes for IG29]
+        //
+        // if (emitComp->opts.disAsm)
+        //{
+        //    emitDispInsAddr(dstRW);
+
+        //    emitDispInsOffs(0, false);
+
+        //    printf("                      %-9s  ; stress-mode injected interrupt\n", "int3");
+        //}
```
Remove?

kunalspathak replied:
I left it commented intentionally so we can quickly uncomment and check it in disassembly. If you still prefer, I can delete it.

BruceForstall reviewed (changes requested):
A few questions

kunalspathak committed:
Review feedback

kunalspathak commented:
> Should PerfScore not count align instructions in an IG following an unconditional branch?

Let me see if it can be done easily.

kunalspathak committed:
Do not include align behind Jmp in PerfScore calculation

kunalspathak commented:
> Let me see if it can be done easily.

Thanks for the suggestion. This gives the real data about PerfScore which is much nicer.

| Name                            | PerfScore diff | Methods Regressed | Methods Improved | Methods NoChange |   |
|---------------------------------|----------------|-------------------|------------------|------------------|---|
| benchmarks.run.windows.x64      | -145.12        | 18                | 184              | 16               |   |
| libraries.pmi.windows.x64       | -270.95        | 31                | 471              | 118              |   |
| aspnet.run.windows.x64          | -440.38        | 6                 | 64               | 7                |   |
| coreclr_tests.pmi.windows.x64   | -94.79         | 12                | 157              | 14               |   |
| benchmarks.run.windows.arm64    | -2273.58       | 0                 | 97               | 25               |   |
| libraries.pmi.windows.arm64     | -1227.70       | 0                 | 346              | 78               |   |
| coreclr_tests.pmi.windows.arm64 | -338.33        | 0                 | 202              | 17               |   |

kunalspathak committed:
jit format and fix a bug

kunalspathak committed:
fix the loopCandidates == 0 scenario

kunalspathak committed:
Add unmarkLoopAlign(), add check for fgFirstBB

kunalspathak committed:
merge conflict fix

kunalspathak committed:
Add missing }

kunalspathak commented:
@BruceForstall  - Can you review it again? I think I have addressed all the feedback.

BruceForstall commented on src/coreclr/jit/block.cpp:
```diff
@@ -1662,3 +1662,22 @@ BBswtDesc::BBswtDesc(Compiler* comp, const BBswtDesc* other)
+//
+void BasicBlock::unmarkLoopAlign(Compiler* compiler DEBUG_ARG(const char* reason))
+{
+    // Make sure we unmark just and count just once.
```
grammar nit:
```suggestion
    // Make sure we unmark and count just once.
```

BruceForstall commented on src/coreclr/jit/compiler.cpp:
```diff
@@ -5208,6 +5215,82 @@ void Compiler::compCompile(void** methodCodePtr, uint32_t* methodCodeSize, JitFl
+    BasicBlock* bbHavingAlign = nullptr;
+    for (BasicBlock* const block : Blocks())
+    {
+        if ((block == fgFirstBB) && block->isLoopAlign())
+        {
+            // Adding align instruction in prolog is not supported
+            // hence skip the align block if it is the first block.
+            loopsToProcess--;
+            continue;
+        }
```
Couldn't we insert a new block before the loop to put the alignment instructions?

kunalspathak replied:
I haven't thought much about it and possible complexities. But definitely, I will explore it.

BruceForstall reviewed (approved):
LGTM. One possible follow-up.

kunalspathak committed:
Grammar nit

Co-authored-by: Bruce Forstall <brucefo@microsoft.com>
